### PR TITLE
Deprecate slow `binary_*` functions in `skimage.morphology`

### PR DESCRIPTION
**Pull request by lagru**
## Description

As discussed in our last hackathon and detailed in our [skimage2 agenda](https://github.com/scikit-image/scikit-image/wiki/API-changes-for-skimage2#proposed-api-changes-doable-independent-of-skimage2:~:text=binary_%20morphological%20functions%20should%20be%20deprecated). 

Most of the time, these functions seem to be slower than our regular gray-type functions, e.g., `ski.morphology.binary_erosion` vs `ski.morphology.erosion`. A potential reason for this might be that the `binary_` funcs use SciPy's (regressed) slow versions under the hood (https://github.com/scipy/scipy/issues/13991).

Also if we add back a faster binary version at some point, can just do it internally for binary input.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
In `skimage.morphology`, deprecate `binary_erosion`, `binary_dilation`, `binary_opening`, 
and `binary_closing` in favor of `erosion`, `dilation`, `opening`, and `closing` 
respectively. The binary versions weren't actually significantly faster than their
non-binary counterparts and sometimes significantly slower. In the future, we might
add optimizations internally to the remaining (general, non-binary) functions for 
when they're used with binary inputs.
```

### CODE REVIEW

**Comment by lagru** [comment]
Here's a quick & dirty script (and it's output on my machine) to get an overview about the performance of the mentioned functions.

<details><summary>Details</summary>

```python
# timeit.py

import timeit
import subprocess

import numpy as np
import skimage as ski

mask_sizes = [10, 1000]
footprint_sizes = [3, 9]
funcs = ["erosion", "dilation", "opening", "closing"]

for func in funcs:
    for mask_size in mask_sizes:
        for footprint_size in footprint_sizes:
            print(f"{mask_size=}, {footprint_size=}")
            setup = (
                "import skimage as ski; import numpy as np;"
                "import skimage.morphology as sm;"
                f"mask = ski.data.binary_blobs(length={mask_size});"
                f"footprint = sm.footprint_rectangle(({footprint_size},) * 2);"
            )
            test_gray = f"sm.{func}(mask, footprint)"
            print("  " + test_gray, end=": ")
            subprocess.run(
                ["python", "-m", "timeit", "-s", setup, test_gray]
            )

            test_binary = f"sm.binary_{func}(mask, footprint)"
            print("  " + test_binary, end=": ")
            subprocess.run(
                ["python", "-m", "timeit", "-s", setup, test_binary]
            )
            print()
```

prints
```
mask_size=10, footprint_size=3
  sm.erosion(mask, footprint): 5000 loops, best of 5: 74.4 usec per loop
  sm.binary_erosion(mask, footprint): 5000 loops, best of 5: 58.5 usec per loop

mask_size=10, footprint_size=9
  sm.erosion(mask, footprint): 5000 loops, best of 5: 70.8 usec per loop
  sm.binary_erosion(mask, footprint): 2000 loops, best of 5: 106 usec per loop

mask_size=1000, footprint_size=3
  sm.erosion(mask, footprint): 50 loops, best of 5: 7.61 msec per loop
  sm.binary_erosion(mask, footprint): 20 loops, best of 5: 11 msec per loop

mask_size=1000, footprint_size=9
  sm.erosion(mask, footprint): 50 loops, best of 5: 7.63 msec per loop
  sm.binary_erosion(mask, footprint): 10 loops, best of 5: 35.2 msec per loop

mask_size=10, footprint_size=3
  sm.dilation(mask, footprint): 5000 loops, best of 5: 75.4 usec per loop
  sm.binary_dilation(mask, footprint): 5000 loops, best of 5: 57.8 usec per loop

mask_size=10, footprint_size=9
  sm.dilation(mask, footprint): 5000 loops, best of 5: 75.8 usec per loop
  sm.binary_dilation(mask, footprint): 2000 loops, best of 5: 116 usec per loop

mask_size=1000, footprint_size=3
  sm.dilation(mask, footprint): 50 loops, best of 5: 7.33 msec per loop
  sm.binary_dilation(mask, footprint): 20 loops, best of 5: 10.9 msec per loop

mask_size=1000, footprint_size=9
  sm.dilation(mask, footprint): 50 loops, best of 5: 7.73 msec per loop
  sm.binary_dilation(mask, footprint): 10 loops, best of 5: 36.4 msec per loop

mask_size=10, footprint_size=3
  sm.opening(mask, footprint): 2000 loops, best of 5: 179 usec per loop
  sm.binary_opening(mask, footprint): 2000 loops, best of 5: 114 usec per loop

mask_size=10, footprint_size=9
  sm.opening(mask, footprint): 2000 loops, best of 5: 184 usec per loop
  sm.binary_opening(mask, footprint): 1000 loops, best of 5: 234 usec per loop

mask_size=1000, footprint_size=3
  sm.opening(mask, footprint): 20 loops, best of 5: 15.4 msec per loop
  sm.binary_opening(mask, footprint): 10 loops, best of 5: 21.9 msec per loop

mask_size=1000, footprint_size=9
  sm.opening(mask, footprint): 20 loops, best of 5: 15.7 msec per loop
  sm.binary_opening(mask, footprint): 5 loops, best of 5: 79.8 msec per loop

mask_size=10, footprint_size=3
  sm.closing(mask, footprint): 2000 loops, best of 5: 177 usec per loop
  sm.binary_closing(mask, footprint): 2000 loops, best of 5: 115 usec per loop

mask_size=10, footprint_size=9
  sm.closing(mask, footprint): 2000 loops, best of 5: 189 usec per loop
  sm.binary_closing(mask, footprint): 1000 loops, best of 5: 247 usec per loop

mask_size=1000, footprint_size=3
  sm.closing(mask, footprint): 20 loops, best of 5: 15.7 msec per loop
  sm.binary_closing(mask, footprint): 10 loops, best of 5: 20.6 msec per loop

mask_size=1000, footprint_size=9
  sm.closing(mask, footprint): 20 loops, best of 5: 14.8 msec per loop
  sm.binary_closing(mask, footprint): 5 loops, best of 5: 71.4 msec per loop
```

</details> 

`binary_*` seem to have a slight advantage for small footprints (3x3) and images (10x10), but are slower, sometimes significantly, for large footprints (9x9) and images (1000x1000).

File: skimage/morphology/tests/test_footprints.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ def test_footprint_dtype(function, args, supports_sequence_decomposition, dtype)
 
 
 @pytest.mark.parametrize("function", ["disk", "ball"])
-@pytest.mark.parametrize("radius", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20, 50, 75, 100])
+@pytest.mark.parametrize("radius", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20])
```
Comment (mkcor) [inline skimage/morphology/tests/test_footprints.py:160]: Is this change unrelated to the rest?

File: skimage/morphology/binary.py [hunk 2]
```diff
@@ -37,6 +40,11 @@ def _iterate_binary_func(binary_func, image, footprint, out, border_value):
 # default with the same dimension as the input image and size 3 along each
 # axis.
 @default_footprint
+@deprecate_func(
+    deprecated_version="0.26",
+    removed_version="0.28",
+    hint="Use `skimage.morphology.erosion` instead.",
```
Comment (mkcor) [inline skimage/morphology/binary.py:46]: Do we care about
https://github.com/scikit-image/scikit-image/blob/96834cf3d57c4a9bad82a471a5548be3ad53a4fa/skimage/morphology/gray.py#L198-L200
(docstring for `skimage.morphology.erosion`) though?

**Review by mkcor — COMMENTED** [review]
Thanks for tackling the skimage2 agenda! :pray: 

There's also an instance of `skimage.morphology.binary_dilation` in
https://github.com/scikit-image/scikit-image/blob/96834cf3d57c4a9bad82a471a5548be3ad53a4fa/skimage/restoration/tests/test_inpaint.py#L9

Do we want to update the `scipy.ndimage.binary_*` as well?

And should we clean up
https://github.com/scikit-image/scikit-image/blob/96834cf3d57c4a9bad82a471a5548be3ad53a4fa/skimage/morphology/misc.py#L17-L26
? Maybe that's implied in the TODO for v0.28?

File: skimage/morphology/tests/test_footprints.py (hunk 1 above)
Comment (lagru) [reply skimage/morphology/tests/test_footprints.py:160]: Yes, see https://github.com/scikit-image/scikit-image/pull/7665/commits/69c82eda3ee4d6c6fa6f96d621575a6d6f8bbbff:

> Remove slow parameters in test_nsphere_series_approximation 
>
> Creating and testing large 3D footprints with ball is very slow so that
the test took 26 s in total on my machine. Removing 50-100 shaves of
~25 s without sacrificing any test coverage I think.

File: skimage/morphology/binary.py (hunk 2 above)
Comment (lagru) [reply skimage/morphology/binary.py:46]: Good catch! I don't see a reason not to mention it and help users avoid weird unexpected edge cases when transitioning.

File: skimage/morphology/tests/test_footprints.py (hunk 1 above)
Comment (mkcor) [reply skimage/morphology/tests/test_footprints.py:160]: "shaves off" (I guess)

Thanks. I admit I didn't review per commit this time!

File: TODO.txt [hunk 3]
```diff
@@ -1,3 +1,8 @@
+Version 0.28
+------------
+* Complete deprecation of `skimage.morphology.binary_*` functions.
```
Comment (mkcor) [inline TODO.txt:3]:
```suggestion
* Complete deprecation of `skimage.morphology.binary_*` functions.
  - Remove reference to `skimage.morphology.binary_erosion` in docstring for `skimage.morphology.erosion`.
  - Remove reference to `skimage.morphology.binary_dilation` in docstring for `skimage.morphology.dilation`. 
```

File: skimage/morphology/binary.py (hunk 2 above)
Comment (mkcor) [inline skimage/morphology/binary.py:46]:
```suggestion
    hint="Use `skimage.morphology.erosion` instead. "
         "Note that, for even-sized footprints, the output will be shifted "
         "by one pixel.",
```

File: skimage/morphology/binary.py [hunk 4]
```diff
@@ -118,6 +126,11 @@ def binary_erosion(image, footprint=None, out=None, *, mode='ignore'):
 
 
 @default_footprint
+@deprecate_func(
+    deprecated_version="0.26",
+    removed_version="0.28",
+    hint="Use `skimage.morphology.dilation` instead.",
```
Comment (mkcor) [inline skimage/morphology/binary.py:132]:
```suggestion
    hint="Use `skimage.morphology.dilation` instead. "
         "Note that, for non-symmetric footprints, `skimage.morphology.dilation` "
         "does not mirror the footprint (unlike "
         "`skimage.morphology.binary_dilation`, which does).",
```

File: skimage/morphology/binary.py (hunk 2 above)
Comment (mkcor) [reply skimage/morphology/binary.py:46]: Exactly. I've expanded the hint message in a review suggestion.

**Comment by lagru** [comment]
> https://github.com/scikit-image/scikit-image/blob/96834cf3d57c4a9bad82a471a5548be3ad53a4fa/skimage/morphology/misc.py#L17-L26

That whole `skimage2ndimage` seems outdated and not to be used anywhere. But for the sake of the scope of this PR I'll only remove the references to the deprecated functions. We can always come back and clean this up later but I don't think it's a notable problem, since it's not really advertised by us anywhere. Long-term / for skimage2 we should make `misc.py` and similar private though I think.

File: skimage/morphology/binary.py (hunk 2 above)
Comment (lagru) [reply skimage/morphology/binary.py]: Thanks, I chose to point the deprecation message at the existing notes in the docstring, see https://github.com/scikit-image/scikit-image/pull/7665/commits/868a8c686b7f40c154cee24fee2fcd9d4f4455ea. That keeps it somewhat shorter, and points to the full explanation. Let me know if you are okay with this.

**Comment by lagru** [comment]
Note the release note in the PR description. This time I chose to expand a bit on the reasoning behind the deprecation since it might otherwise look like a unnecessary deprecation.

**Comment by mkcor** [comment]
> Note the release note in the PR description. This time I chose to expand a bit on the reasoning behind the deprecation since it might otherwise look like a unnecessary deprecation.

Thanks for this thorough release note, indeed! I would just add Oxford commas (before "and `binary_closing`" and before "and `closing`"). And another comma before "respectively," but here it's me being picky!

I would replace "to the remaining non-deprecated functions" with "to the remaining (general, non-binary) functions for when they're used with binary inputs."

File: benchmarks/benchmark_morphology.py [hunk 5]
```diff
@@ -48,12 +48,28 @@ def peakmem_skeletonize(self):
...
+    # skip rectangle as roughly equivalent to square
+    param_names = ["shape", "radius"]
+    params = [
+        ((512, 512),),
+        (1, 3, 5, 15, 25, 40),
+    ]
 
+    def setup(self, shape, radius):
+        rng = np.random.default_rng(123)
+        # Make an image that is mostly True, with random isolated False areas
```
Comment (mkcor) [inline benchmarks/benchmark_morphology.py:61]:
```suggestion
        # Create an image that is mostly True, with random isolated False areas
```

File: skimage/morphology/isotropic.py [hunk 6]
```diff
@@ -9,7 +9,7 @@
 def isotropic_erosion(image, radius, out=None, spacing=None):
     """Return binary morphological erosion of an image.
 
-    This function returns the same result as :func:`skimage.morphology.binary_erosion`
+    This function returns the same result as :func:`skimage.morphology.erosion`
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:12]: If it actually returns the same result as `binary_erosion`, when changing to `erosion`, I guess we need to include the same warning that differentiates `erosion` from `binary_erosion`, i.e., "Note the pixel shift by 1 for even-sized footprints (see docstring notes)."

File: skimage/morphology/isotropic.py [hunk 7]
```diff
@@ -57,7 +57,7 @@ def isotropic_erosion(image, radius, out=None, spacing=None):
 def isotropic_dilation(image, radius, out=None, spacing=None):
     """Return binary morphological dilation of an image.
 
-    This function returns the same result as :func:`skimage.morphology.binary_dilation`
+    This function returns the same result as :func:`skimage.morphology.dilation`
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:60]: If it actually returns the same result as `binary_dilation`, when changing to `dilation`, I guess we need to include the same warning that differentiates `dilation` from `binary_dilation`, i.e., "Note the lack of mirroring for non-symmetric footprints (see docstring notes)."

File: skimage/morphology/isotropic.py (hunk 6 above)
Comment (mkcor) [reply skimage/morphology/isotropic.py:12]: Maybe, after the 'Returns' section:
```py
    Notes                                                                      
    -----                                                                     
    For even-sized footprints, :func:`skimage.morphology.erosion` and
    this function produce an output that differs: one is shifted by one pixel
    compared to the other. :func:`skimage.morphology.pad_footprint´ is available
    to account for this.
```
?

File: skimage/morphology/isotropic.py (hunk 7 above)
Comment (mkcor) [reply skimage/morphology/isotropic.py:60]: Maybe, after the 'Returns' section:
```py
    Notes                                                                      
    -----
    For non-symmetric footprints, :func:`skimage.morphology.dilation`
    and this function produce an output that differs:
    This function mirrors the footprint, whereas `dilation` does not.
    :func:`skimage.morphology.mirror_footprint` is available to correct for this.
```

File: skimage/morphology/isotropic.py (hunk 6 above)
Comment (lagru) [reply skimage/morphology/isotropic.py:12]: Hmm, I'm not sure that the docstrings claim "the same result" makes a lot of sense in the first place. This function takes a `radius` to determine the neighborhood while both other `erosion` functions take in a `footprint`. So `isotropic_erosion` seems to be limited to a circular footprint of uneven size!

So we should probably reword this. What do you think?

File: skimage/morphology/isotropic.py (hunk 7 above)
Comment (lagru) [reply skimage/morphology/isotropic.py:60]: See general discussion in https://github.com/scikit-image/scikit-image/pull/7665#discussion_r1929768123.

File: skimage/morphology/isotropic.py (hunk 6 above)
Comment (lagru) [reply skimage/morphology/isotropic.py]: See https://github.com/scikit-image/scikit-image/pull/7665/commits/549dc4732aee767b83f02ab49deb1c4595574c13 as a sugggestion.

File: skimage/morphology/isotropic.py [hunk 8]
```diff
@@ -9,8 +9,9 @@
 def isotropic_erosion(image, radius, out=None, spacing=None):
     """Return binary morphological erosion of an image.
 
-    This function returns the same result as :func:`skimage.morphology.binary_erosion`
-    but performs faster for large circular structuring elements.
+    Compared to the more general :func:`skimage.morphology.erosion`, this
+    function only supports binary input and circular neighborhoods.
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:13]:
```suggestion
    function only supports binary inputs and circular footprints.
```
It seems that, in the rest of the codebase, we use mostly the term 'footprint' and, sometimes, 'structuring element.'

File: skimage/morphology/isotropic.py [hunk 9]
```diff
@@ -9,8 +9,9 @@
...
+    However, it may perform faster for large neighborhoods.
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:14]:
```suggestion
    However, it performs typically faster for large (circular) footprints.
```

File: skimage/morphology/isotropic.py [hunk 10]
```diff
@@ -20,7 +21,7 @@ def isotropic_erosion(image, radius, out=None, spacing=None):
     image : ndarray
         Binary input image.
     radius : float
-        The radius by which regions should be eroded.
+        The radius defining the neighborhood of the operation.
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:24]:
```suggestion
        The radius of the footprint.
```

File: skimage/morphology/isotropic.py [hunk 11]
```diff
@@ -57,8 +77,9 @@ def isotropic_erosion(image, radius, out=None, spacing=None):
 def isotropic_dilation(image, radius, out=None, spacing=None):
     """Return binary morphological dilation of an image.
 
-    This function returns the same result as :func:`skimage.morphology.binary_dilation`
-    but performs faster for large circular structuring elements.
+    Compared to the more general :func:`skimage.morphology.dilation`, this
+    function only supports binary input and circular neighborhoods.
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:81]: I would make equivalent suggestions here.

File: skimage/morphology/isotropic.py [hunk 12]
```diff
@@ -68,7 +89,7 @@ def isotropic_dilation(image, radius, out=None, spacing=None):
     image : ndarray
         Binary input image.
     radius : float
-        The radius by which regions should be dilated.
+        The radius defining the neighborhood of the operation.
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:92]:
```suggestion
        The radius of the footprint.
```

File: skimage/morphology/isotropic.py [hunk 13]
```diff
@@ -142,6 +183,25 @@ def isotropic_opening(image, radius, out=None, spacing=None):
         and thresholding of distance maps, Pattern Recognition Letters,
         Volume 13, Issue 3, 1992, Pages 161-166.
         :DOI:`10.1016/0167-8655(92)90055-5`
+
+    Examples
+    --------
+    Remove undesired connection between two bright regions
```
Comment (mkcor) [inline skimage/morphology/isotropic.py:189]:
```suggestion
    Remove connection between two bright regions
```

**Review by mkcor — COMMENTED** [review]
Thanks, @lagru!

5776eb27753adeff7a is a very nice addition, but I'm missing what's going on for the `isotropic_opening`  and `isotropic_closing`  examples, since `np.testing.assert_array_equal(image, result)` (no change?).

File: skimage/morphology/isotropic.py (hunk 6 above)
Comment (mkcor) [reply skimage/morphology/isotropic.py]: You're right; I skimmed instead of glancing over this... I very much like 549dc4732aee767b83f02ab49deb1c4595574c13; thanks for this. I just made a few suggestions (latest review). Also, when we merge, I'd like to remember to change "uneven size" into "odd size" in the PR description.

File: skimage/morphology/isotropic.py (hunk 8 above)
Comment (lagru) [reply skimage/morphology/isotropic.py:13]: Agreed but I wasn't actually sure if `footprint` would make sense in this case since the function doesn't actually allow passing a `footprint` arg. Instead you define the footprint with a radius. I think "neighborhood" is a term that we are using [elsewhere](https://scikit-image.org/docs/stable/api/skimage.morphology.html#skimage.morphology.closing) and seems to fit better IMO. What do you think?

**Comment by lagru** [comment]
I addressed all your comments @mkcor. Should be ready to go. :D

File: skimage/morphology/isotropic.py (hunk 8 above)
Comment (mkcor) [reply skimage/morphology/isotropic.py]: The term 'neighborhood' is more general; in the case of these morphological operations, we talk about 'footprints' or 'structuring elements' (more specifically).

Aww, it seems that the `skimage.morphology.closing` docstring would need some love ('“close” up' instead of '“close up”' and, more importantly, the description for `footprint` isn't great: You may say that the footprint is a kind of neighborhood, but it's weird to start with "THE neighborhood" and then use another term; *and*, let's check, but it shouldn't necessarily be *2D*!!).

**Review by mkcor — APPROVED** [review]
:tada:

**Comment by lagru** [comment]
Converting to draft so this is not accidentally merged into the pending 0.25.2 patch release.

**Review by jarrodmillman — APPROVED** [review]
